File: src/sw/sa_pipeline/stage8_sa.py
class Stage8SA:
    """
    Eighth Pipe from SA_Verilog. This pipe is responsible to execute the 2-2 
    additions for the distances found in the left pipe.
    """

    def __init__(self):
        """

        """
        self.new_output: dict = {
            'th_idx': 0,
            'th_valid': False,
            'dc': 0,
            'ds': 0
        }
        self.old_output: dict = self.new_output.copy()

    def compute(self, st7_input: dict):
        """

        @param st7_input:
        """
        # moving forward the ready outputs
        self.old_output = self.new_output.copy()

        st7_th_idx: int = st7_input['th_idx']
        st7_th_valid: bool = st7_input['th_valid']
        st7_dc: int = st7_input['dc']

        st7_dvas: int = st7_input['dvas']
        st7_dvbs: int = st7_input['dvbs']

        ds: int = st7_dvas + st7_dvbs

        self.new_output: dict = {
            'th_idx': st7_th_idx,
            'th_valid': st7_th_valid,
            'dc': st7_dc,
            'ds': ds
        }

File: src/cpp/sw/sa_pipeline/stage8_sa.cpp
#include <iostream>

class Stage8SA {
private:
    int* new_output;
    int* old_output;

public:
    Stage8SA() {
        new_output = new int[4]();
        old_output = new int[4]();
    }

    ~Stage8SA() {
        delete[] new_output;
        delete[] old_output;
    }

    void compute(int* st7_input) {
        std::copy(new_output, new_output + 4, old_output);

        int st7_th_idx = st7_input[0];
        bool st7_th_valid = st7_input[1];
        int st7_dc = st7_input[2];

        // fixme only for debugging
        // if st7_th_idx == 0:
        //    z = 1

        int st7_dvas = st7_input[3];
        int st7_dvbs = st7_input[4];

        int ds = st7_dvas + st7_dvbs;

        std::copy(st7_input, st7_input + 4, new_output);
        new_output[2] = st7_dc;
        new_output[3] = ds;
    }
};